File: sdk/python/src/tests/provenance_test.py
from fraise_sdk.models import Hit, RecallResult
from fraise_sdk.query import build_remember


def test_build_remember_quotes_source_reference():
    assert build_remember("fact", graph=2, source="Tool Call / Nick's session") == "remember@2 'fact' source:'Tool Call / Nick''s session'"


def test_hit_parses_provenance_and_contributions():
    hit = Hit.from_json({"value": "fact", "score": 0.75, "source": "session:abc", "contributions": [{"source": "graph", "score": 2, "rank": 1, "count": 2, "via": "weather", "degree": 3}]})
    assert hit.source == "session:abc"
    assert hit.contributions[0].via == "weather"
    assert hit.contributions[0].degree == 3


def test_old_hit_payload_remains_compatible():
    hit = Hit.from_json({"value": "fact", "score": 1.0})
    assert hit.source is None
    assert hit.contributions == []


def test_recall_result_parses_explain_background():
    assert RecallResult.from_json({"count": 0, "hits": [], "background": 0.125}).background == 0.125

File: internal/query/provenance_test.go
package query

import (
	"encoding/json"
	"strings"
	"testing"
	"time"

	"github.com/RonsenbergVI/fraise/internal/graph"
)

func TestRememberHashDistinguishesSourceWithoutChangingLegacyEmptyHash(t *testing.T) {
	base := Remember[string, float32]{Value: "fact"}
	a, b := base, base
	a.Source, b.Source = "session:a", "session:b"
	if a.Hash(&fakeHasher{}) == b.Hash(&fakeHasher{}) {
		t.Fatal("different provenance references shared a plan-cache key")
	}
	if legacy := base.Hash(&fakeHasher{}); strings.Contains(legacy, "|src=") {
		t.Fatalf("empty provenance changed legacy cache identity: %q", legacy)
	}
}

func TestHitSourceAppearsOnlyInExplainMode(t *testing.T) {
	var node graph.Node[string] = graph.Fact[string]{
		NodeAttributes: graph.NodeAttributes{Value: "fact", Timestamp: time.Unix(1, 0).UTC()},
		Source:         "session:abc",
	}
	plain, err := json.Marshal(Hit[string, float32]{Node: &node, Score: 1})
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(plain), `"source"`) {
		t.Fatalf("plain recall leaked provenance: %s", plain)
	}
	explained, err := json.Marshal(Hit[string, float32]{
		Node:          &node,
		Score:         1,
		Source:        "session:abc",
		Contributions: []HitContribution[float32]{{Source: "text", Score: 1, Rank: 0, Count: 1}},
	})
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(string(explained), `"source":"session:abc"`) {
		t.Fatalf("explained recall omitted provenance: %s", explained)
	}
}
